Add OAuth2 authentication provider

hussainweb opened the pull request:
This change depends on the changes in PR #940. It uses the new method introduced in that PR
to identify the resource being requested and get the plugin definition to read the oauth2
server and scope to use for authorizing the request.

This would probably need to be changed once #940 is merged or a better method to access the
request being processed is identifed (as mentioned in #939). I am creating this PR in hope
of clarifying the changes in PR #940.

hussainweb committed:
Add OAuth2 authentication provider

This change depends on the changes in PR #940. It uses the new method introduced in that PR
to identify the resource being requested and get the plugin definition to read the oauth2
server and scope to use for authorizing the request.

This would probably need to be changed once #940 is merged or a better method to access the
request being processed is identifed (as mentioned in #939). I am creating this PR in hope
of clarifying the changes in PR #940.

e0ipso commented on src/Plugin/authentication/OAuth2ServerAuthentication.php:
```diff
@@ -0,0 +1,74 @@
+ * Authentication support for oauth2_server.
+ *
+ * @Authentication(
+ *   id = "oauth2_auth",
```
`oauth2_auth` ➡️  `oauth2`

hussainweb replied:
I was using `basic_auth` as an example. I see other plugins don't follow this convention. Changing...

e0ipso commented on src/Plugin/authentication/OAuth2ServerAuthentication.php:
```diff
@@ -0,0 +1,74 @@
+   * {@inheritdoc}
+   */
+  public function applies(RequestInterface $request) {
+    return module_exists('oauth2_server') && $this->getResourcePluginIdFromRequest();
```
This method should check the request to see if it contains the necessary info to perform the required checks.

This probably means moving `$oauth2_info = $this->getOAuth2Info($request);` here.

e0ipso commented on src/Plugin/authentication/OAuth2ServerAuthentication.php:
```diff
@@ -0,0 +1,74 @@
+    }
+  }
+
+//  protected function getOAuth2Info() {
```
Leftovers.

hussainweb replied:
@e0ipso, Not really leftover, but until #940 goes in, I was keeping in an alternate version I could uncomment for testing. I will remove it once you approve the direction.

e0ipso commented on src/Plugin/authentication/OAuth2ServerAuthentication.php:
```diff
@@ -0,0 +1,74 @@
+    if ($result instanceof \OAuth2\Response) {
+      throw new UnauthorizedException($result->getResponseBody(), $result->getStatusCode());
+    }
+    elseif (is_array($result) && !empty($result['user_id'])) {
```
Make this whole `ifelse`:

``` php
if (empty($result['user_id'])) {
  return NULL;
}
return user_load($result['user_id']);
```

hussainweb replied:
I am not very sure. I thought there might be something useful in checking for `\OAuth2\Response` and handling that?

e0ipso replied:
For clarity, I meant «Make this whole `elseif`:»

I'd leave the `if` above alone, but simplify the logic (without leaving any execution path without a `return`).

e0ipso commented on src/Plugin/authentication/OAuth2ServerAuthentication.php:
```diff
@@ -0,0 +1,74 @@
+//    return [variable_get('oauth2_server_restful_server'), variable_get('oauth2_server_restful_scope')];
+//  }
+
+  protected function getOAuth2Info($request) {
```
Needs docblock.

e0ipso commented on src/Plugin/authentication/OAuth2ServerAuthentication.php:
```diff
@@ -0,0 +1,74 @@
+
+  protected function getOAuth2Info($request) {
+    $plugin_id = $this->getResourcePluginIdFromRequest();
+    $plugin = ResourcePluginManager::create('cache', $request)->getDefinition($plugin_id);
```
`$plugin` ➡️  `$plugin_definition`.

hussainweb replied:
I did use `->getPlugin` in the beginning. It results in an infinite recursion. Calling `getPlugin` in `authenticate` or `applies` ultimately tries to create the plugin again, which tries to run authentication again. That is why I created #940.

e0ipso commented on src/Plugin/authentication/OAuth2ServerAuthentication.php:
```diff
@@ -0,0 +1,74 @@
+    $plugin_id = $this->getResourcePluginIdFromRequest();
+    $plugin = ResourcePluginManager::create('cache', $request)->getDefinition($plugin_id);
+
+    $server = !empty($plugin['oauth2Server']) ? $plugin['oauth2Server'] : variable_get('oauth2_server_restful_server');
```
Let's drop the global variables: `oauth2_server_restful_server` and `oauth2_server_restful_scope` and require the annotation to contain the info. If the info is not there we can throw a 5XX exception.

hussainweb replied:
The `options` key is on `@Authentication` annotation. I must put that on the OAuth2Authentcation class. That does not help the use case where resources would want to specify the server and scope.

e0ipso commented on src/Plugin/authentication/OAuth2ServerAuthentication.php:
```diff
@@ -0,0 +1,74 @@
+    $plugin_id = $this->getResourcePluginIdFromRequest();
+    $plugin = ResourcePluginManager::create('cache', $request)->getDefinition($plugin_id);
+
+    $server = !empty($plugin['oauth2Server']) ? $plugin['oauth2Server'] : variable_get('oauth2_server_restful_server');
```
And let's move it under the `'options'` key in the annotation. See https://github.com/RESTful-Drupal/restful/blob/7.x-2.x/src/Annotation/Authentication.php#L35-L40

e0ipso commented on src/Plugin/authentication/OAuth2ServerAuthentication.php:
```diff
@@ -0,0 +1,74 @@
+
+  protected function getOAuth2Info($request) {
+    $plugin_id = $this->getResourcePluginIdFromRequest();
+    $plugin = ResourcePluginManager::create('cache', $request)->getDefinition($plugin_id);
```
Also, did you try using the resource manager's: `->getPlugin($plugin_id, $request)` instead?

e0ipso replied:
If you do that you'll be able to leverage https://github.com/RESTful-Drupal/restful/pull/942/files#r76750941 here as well.

e0ipso commented on src/Plugin/authentication/OAuth2ServerAuthentication.php:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  protected function getResourcePluginIdFromRequest() {
+    $resource_name = restful()->getResourceManager()->getResourceIdFromRequest();
```
Add a protected attribute in the class `$resourceManager` and initialize it in this class' constructor to `restful()->getResourceManager()`. That way there is a single instance used across the class.

hussainweb replied:
It was a simple getter, so I thought it would always be the same instance. But I see your point.

e0ipso commented:
Did you try that this works well with nested resources?

Imagine that the user can only view articles, but not user profiles. Would `/articles/123?include=author` fail accordingly to show the profile in the response?

This resource embedding clones part of the request. We need to make sure that permissions are still enforced.

e0ipso commented:
@hussainweb thank you for working on this!! I think that this is a very important feature.

I added a couple of comments above.

hussainweb commented:
@e0ipso, thank you for all the comments and an elaborate review. I am pushing changes now.

> Imagine that the user can only view articles, but not user profiles. Would /articles/123?include=author fail accordingly to show the profile in the response?

I am not very sure. If the request in resource manager is replaced for the sub-request (which I suspect), then yes. It should work as expected. I don't have any resources right now which embed this partially. Any suggestions on how to easily test this?

Also, how do other authentication types enforce this? e.g., BasicAuthentication also reads the code from the request, and if that works for sub-requests, this should too.

hussainweb committed:
Fix OAuth2 authentication plugin based on review

hussainweb committed:
Avoid loading the plugin definition for all requests

The OAuth2 authentication plugin tries to load the plugin as part of `applies()` check, even
if a plugin id cannot be determined. This change makes sure there is a plugin id before
trying to load.

e0ipso commented on src/Plugin/authentication/OAuth2ServerAuthentication.php:
```diff
@@ -0,0 +1,107 @@
+      return NULL;
+    }
+
+    $scope = !empty($plugin_definition['oauth2Scope']) ? $plugin_definition['oauth2Scope'] : variable_get('oauth2_server_restful_scope');
```
Can we get rid of the `variable_get`?

e0ipso replied:
(applies to other places as well)

hussainweb committed:
Remove variable_get from OAuth2 authentication plugin

e0ipso commented:
This looks good to me 👍. Thanks for this killer feature.

I'd love to see a section to the [wiki](https://github.com/RESTful-Drupal/restful/wiki/3.-Authentication) on how to configure this to have a resource use this authentication method. That will empower users even more.

hussainweb commented:
I'll try to get around to the wiki this week. Thank you for merging this!
